browser/auto_import_muon.cc

```diff
@@ -0,0 +1,129 @@
+  }
+
+  bool ended() const {
+    return ended_;
+  }
+
+ private:
+  // Set if the import has ended.
+  bool ended_;
+
+  base::Closure callback_for_import_end_;
+
+  DISALLOW_COPY_AND_ASSIGN(ImportEndedObserver);
+};
+
+void ImportFromSourceProfile(const importer::SourceProfile& source_profile,
+                             Profile* target_profile,
+                             uint16_t items_to_import) {
+  // Deletes itself
+  ExternalProcessImporterHost* importer_host =
+      new ExternalProcessImporterHost;
+  // Don't show the warning dialog if import fails
+  importer_host->set_headless();
+
+  ImportEndedObserver observer;
+  importer_host->set_observer(&observer);
+  importer_host->StartImportSettings(source_profile,
+                                     target_profile,
+                                     items_to_import,
+                                     new ProfileWriter(target_profile));
+  // If the import process has not errored out, block on it.
+  if (!observer.ended()) {
+    base::RunLoop loop;
+    observer.set_callback_for_import_end(loop.QuitClosure());
+    loop.Run();
+    observer.set_callback_for_import_end(base::Closure());
+  }
+}
+
+void AutoImportMuon() {
+  LOG(INFO) << "In brave::AutoImportMuon";
+  base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+  if (!command_line.HasSwitch(switches::kUpgradeFromMuon))
+    return;
+
+  LOG(INFO) << "Starting auto-import from Muon";
+
+  // ImporterList::DetectSourceProfiles
+  // -> profiles_loaded_callback
+  // based on https://cs.chromium.org/chromium/src/chrome/browser/first_run/first_run.cc?l=560-577&rcl=e2569db721bb3bd6b12991d581c60933bf555f30
+  base::RunLoop run_loop;
+  auto importer_list = std::make_unique<ImporterList>();
+  importer_list->DetectSourceProfiles(
+      g_browser_process->GetApplicationLocale(),
+      false,  // include_interactive_profiles
+      run_loop.QuitClosure());
+  run_loop.Run();
+
+  bool brave_profile_found = false;
+  size_t brave_profile_index = 0;
+  for (size_t i = 0; i < importer_list->count(); i++) {
+    const auto& source_profile = importer_list->GetSourceProfileAt(i);
+    if (source_profile.importer_type == importer::TYPE_BRAVE) {
+      brave_profile_found = true;
+      brave_profile_index = i;
+      break;
+    }
+  }
+  if (!brave_profile_found) {
+    LOG(INFO) << "Could not find Muon profile to auto-import";
+    return;
+  }
+
+  const importer::SourceProfile& source_profile =
+      importer_list->GetSourceProfileAt(brave_profile_index);
+
+  // Import every possible type of data from the Muon profile
+  uint16_t items_to_import = 0;
+  items_to_import |= source_profile.services_supported;
+
+  // Gross, but good enough for government work? Seems to be a common
+  // part throughout the rest of the Brave codebase.
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  Profile* target_profile = profile_manager->GetLastUsedProfile();
```

Comment:
this should return the same profile, but might be safer to use `profile_manager->GetProfile(profile_manager->GetInitialProfileDir())` because that will definitely return the same profile created by `ProfileManager::CreateInitialProfile`

Reply:
This is weird. `profile_manager->GetLastUsedProfile` works, but we should use something more explicit if we can; however:

- `profile_manager->GetProfileByPath(profile_manager->GetInitialProfileDIr())` fails because `GetInitialProfileDir` returns nullptr.
- `profile_manager->GetProfile(profile_manager->GetInitialProfileDIr())` will create the profile if it doesn't already exist; unfortunately, using it causes a DCHECK failure.

Reply:
Sorry, initial profile creation should be sync (see [CreatePrimaryProfile](https://cs.chromium.org/chromium/src/chrome/browser/chrome_browser_main.cc?l=1557-1559&rcl=8b3c4223c6684dddccaea99ad06331e3738da7b2)), but if that's the case then I don't understand why GetProfileByPath would ever return nullptr when called from PreMainMessageLoopRun.

Reply:
that is strange, but `profile_manager->GetLastUsedProfile` works? If so just go with that. At that point there shouldn't be any other profiles that we have to worry about ending up in LastUsed before we make the call

Reply:
`GetLastUsedProfile` has worked consistently in my (very limited) testing. I agree we should go with it for now, although I wonder whether there is a race with profile creation on the very first run of brave-core; if so, `GetLastUsedProfile` might fail intermittently.